mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp

```diff
@@ -0,0 +1,1029 @@
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements a loop interchange pass that optimizes for locality
+// (spatial and temporal - both self and group) and parallelism for multicores,
+// so as to minimize the frequence of synchronization. The pass works for both
+// perfectly nested and implerfectly nested loops (any level of nesting). However
+// in presence of affine.if statements and/or non-rectangular iteration space,
+// the pass simply bails out - leaving the original loop nest unchanged.
+// The pass is triggered by the command line flag -affine-loop-interchange.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetail.h"
+#include "mlir/Analysis/AffineAnalysis.h"
+#include "mlir/Analysis/Utils.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Affine/Passes.h"
+#include "mlir/Transforms/LoopUtils.h"
+#include "llvm/ADT/SmallSet.h"
+#include "llvm/ADT/SmallVector.h"
+#include <algorithm>
+#include <cmath>
+
+using namespace mlir;
+
+namespace {
+struct LoopInterchange : public AffineLoopInterchangeBase<LoopInterchange> {
+  void runOnFunction() override;
+  void handleImperfectlyNestedAffineLoops(Operation &funcOp);
+
+protected:
+  void getAffineAccessMatrices(
+      AffineForOp &rootForOp,
+      SmallVector<Operation*,8> &loadAndStoreOps,
+      DenseMap<Value, int64_t> &loopVarIdMap,
+      unsigned AffineForOpLoopNestSize
+      );
+
+  void getAffineLoopCarriedDependencies(AffineForOp &rootForOp,
+      ArrayRef<Operation*> loadAndStoreOps,
+      unsigned maxLoopDepth);
+
+  void getElementsSize(AffineForOp &rootForOp);
+
+  void buildReferenceGroups(AffineForOp innermostForOp, 
+      SmallVector<Operation*,8> &loadAndStoreOps,
+      unsigned maxDepth, unsigned forOpLevelInOriginalLoopNest);
+  
+  void getCacheLineAccessCounts(
+      ArrayRef<AffineForOp> AffineLoopNest,
+      SmallVector<Operation*,8>&loadAndStoreOps);
+  
+  long double getTemporalReuseCost(ArrayRef<unsigned> permutation);
+  
+  SmallVector<unsigned,0> getBestPermutation(
+      SmallVector<AffineForOp,0> &AffineLoopNest,
+      DenseMap<Value, int64_t>&loopVarIdMap);
+  
+  void clear(SmallVector<AffineForOp, 0> &AffineForOpLoopNest,
+             DenseMap<Value, int64_t> &loopVarIdMap, 
+      int *loopVarIdCount);
+
+  constexpr static int cache_line_size = 64;
+
+  // For each MemRef access function Ax+B, it is
+  // the collection of A's indexed by respective
+  // load/store operation.
+  DenseMap<Operation *, SmallVector<std::vector<int64_t>, 0>>
+      loopAccessMatrices;
+
+  // For a MemRef access function Ax+B, this represents B.
```

Comment:
All of the member variable comments should be triple `///` doc comments.

Reply:
Done.